### ISSUE TITLE
Lack of compiler API cancellability is a concern.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#4931]
The current compiler API is mostly non-cancellable.  While some high level, known-expensive, API points are cancellable, most API points are not.  These entry points could cause expensive IO to happen deep under the covers (esp. as nearly any compiler API may cause metadata reading to happen).  

We knew about this issue with Roslyn v1, but considered it acceptable.  While the IO might be expensive, it would traditionally be on a background thread, and so not too bad (though still not great).

@tmat has made me aware of some future work that can highly exacerbate this problem.  Namely that metadata reference reading can involve NuGet, which means they may involve the network.  Network calls greatly exacerbate the problem and really must be fixed.  

This is a problem both for background threads and for foreground threads.  While backgrounds threads (mostly) will not impact the user, this can put a negative burden on our threadpool, which affects all the work we do.   For the foreground thread this can provide a massively bad experience for the user.  For example, consider a simple foreground feature (like goto-def):
1. The user invokes the feature.
2. We pop up a threaded wait dialog and start doing all our processing.
3. We call into the compiler.  The compiler needs to make a network call for NuGet.
4.  This takes forever (for any reasonable network reason).  The user tries to cancel.
5. Nothing happens.  There is no path for this cancellation information to reach the appropriate compiler layer.
6. User is stuck waiting for an indeterminate amount of time.  The only way to unblock may be to kill VS.

This is also an issue even if the feature is not on the UI thread.  For example, say we had some sort of 'async' non-ui thread goto-def feature.  The above then becomes:
1. The user invokes the feature.
2. We call into the compiler.  The compiler needs to make a network call for NuGet.
3.  This takes forever (for any reasonable network reason).  
4. Features are impacted because we're starving the threadpool threads on all this IO.

There are different solutions we can take here, with varying levels of attractiveness.

The first would be to simply augment every compiler semantic API with a cancellable alternative, and move the entire LS over to using this.  This is not a very attractive path forward.   The entire compiler semantic API would be effectively duplicated, and we'd have to update an enormous swath of code.  Simple code like `method.Parameters[0].Type.ContainingType` would become `method.GetParameter(cancellationToken).GetAt(0, cancellationToken).GetType(cancellationToken).GetContainingType(cancellationToken)`.  _Shudder_.

Another alternative, raised by tomas, would be to have some sort of "Task local context" (possibly  [CallContext](https://msdn.microsoft.com/en-us/library/system.runtime.remoting.messaging.callcontext%28v=vs.110%29.aspx)) that features would set before making lots of call into the compiler. This task local context would be used to store the cancellation token in an ambient fashion, that the compiler layers could then query for deep down when making expensive IO calls (from disk or over the network).

This approach has the potential to keep our IDE and compiler-interaction code very clean, while also ensuring that deep compiler layers are aware of when the IDE layer wants them to cancel so we can clean up quickly and allow the thread pool thread we're using to get onto other more important work.  

IDE team, what do you think about this issue?

**Comment by CyrusNajmabadi** [comment]
@balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie @jasonmalinowski @rchande @dpoeschl 

Feedback welcome.

**Comment by CyrusNajmabadi** [comment]
@gafter @jaredpar @AlekseyTs @VSadov as well.

**Comment by Pilchie** [comment]
also tagging @mattwar

**Comment by tmat** [comment]
@KevinH-MS

**Comment by CyrusNajmabadi** [comment]
Note: if we did follow the "ambient cancellation" model, we could potentially consider using it for the IDE layers as well.  This would make a bunch of internal APIs far more attractive, and would make a lot of our own code much cleaner.  For public entrypoints we'd still follow the existing 'takes cancellation token' pattern.  But internally we could then switch to ambientness.

One thing we have to be careful of is existing codepaths that change the cancellation token, or which supress cancellation (which is effectively just changing the token to ```CancellationToken.None```).  We'd need to be super careful that we were properly overriding and restoring the cancellation tokens appropriately.

**Comment by CyrusNajmabadi** [comment]
Note: this also sounds a bit like F# (which i  believe has ambient cancellation tokens for their workflows).   I'm adding @vladima to see if he has some insight on the feasibility of this approach, as well as any potential pitfalls for us to be aware of.  @vladima, if there's any other experts on this topic that you think would be worth talking to, let me know.

Thanks!

**Comment by gafter** [comment]
The concept of "ambient state" is a design smell. I am also concerned that supporting this "properly" could add a lot to the expense of running the compiler, as there are many places where we depend for correctness on the fact that an operation we start is completed.

I suggest you stop invoking the compiler on a foreground thread.

**Comment by CyrusNajmabadi** [comment]
@gafter We can stop invoking the compiler on a foreground thread.  The problem still remains that it may perform deeply expensive operations in an non-cancellable manner.  These operations consume significant resources and can affect the throughput and other performance of the system.

For example, on background threads, we may have N different operations calling into the compiler for basic (non-cancellable) properties.  All of these may end up blocking while the compilation layer tries to do the nugget resolution.  Because of this, the threadpool ends up totally starved.

Ideally, long running, blocking operations, would be transformed into cancellable async methods.  Of course, that's just as invasive and impactful on the end API.  

The hope is that there's some way for the compiler to keep it's *apparent* API simplicity, but also ensure that it doesn't behave in a manner that is detrimental to the layers that sit above it.  Right now the IO it performs is already bad in that regard, and it sounds like it will get worse in the future.   This is an issue regardless of which threads we're actually using to call into the compiler.

**Comment by mattwar** [comment]
I don't think the compiler API should be doing any nuget magic w/o explicit API that is cancellable.  We should reconsider how this work is being done.

**Comment by CyrusNajmabadi** [comment]
Another potential option might be to ask for a compilation with a "swapped out metadata reader".  i.e. you get the normal compilation through some mechanism, and you say "compilation.WithCancellableMetadataReader(cancellationToken)" giving back a new compiler instance that shares most state with the compilation it was forked off of, but which also threads this cancellation token you provided into any internal calls it makes into IO operations.

@tmat Would such a thing be possible?

**Comment by CyrusNajmabadi** [comment]
@mattwar Note: the issue exists even without the Nuget wrinkle.  Nuget just makes it very clear how bad the issue is.  Currently we perform non-cancellable IO.  That's something which is normally considered to be a very bad no-no.  Ideally we can solve that issue at the same time.  This should be very desirable for all systems that are using the compiler layer and ever do anything that wants to cancel.  

The build-server comes to mind as another potential client of this.

**Comment by gafter** [comment]
@CyrusNajmabadi I'm not sure how much complexity we are willing to pay for in the compiler to allow cancellation essentially anywhere. There are many points in the compiler where cancellation would leave the compiler in an inconsistent internal state.

**Comment by CyrusNajmabadi** [comment]
@gafter 

That's definitely an interesting concern.  @tmat do you thoughts on this problem?

**Comment by tmat** [comment]
The WithResolvedStuff(CancellationToken) is equivalent to the original proposal to add cancellation token to CompilationOptions. 

We just need to make sure that the original compilation and the new one share as much state as possible, so that multiple services that clone the compilation with different tokens can share as much data as possible. The metadata reference binding is already implemented that way. @KevinH-MS is moving syntax trees and decl table to the same model to facilitate #load. 

We would need to change the rest of the significant compilation state (diagnostics, global namespace, etc.) to be shareable across compilations as well.

**Comment by tmat** [comment]
Oh, and there is a problem. AssemblySymbol is coupled with Compilation :(, we can't share that.

**Comment by gafter** [comment]
We cannot share symbols across compilations. Metadata references, syntax trees, and the decl table, yes. But not (source) symbols.

Processing the code to resolve semantics can be computationally expensive. We can thread cancellation tokens fairly shallowly into some calls on Compilation (e.g. when you ask for diagnostics or call emit), but beyond a certain depth I don’t think we want to thread into the bowels of, for example, overload resolution. Writing the output files is something that the caller can control (the portable compiler layer just dumps data into a stream the caller provides). But APIs such as `TypeSymbol.BaseClass` and `LocalSymbol.Type` may sometimes involve reading metadata, and that is far too deep in the guts of the compiler for every intermediate call on the stack to be prepared for cancellation. I believe when metadata reading fails we just treat a symbol we get back as a bad symbol – and it will be bad forever thereafter.

Perhaps a fairly shallow threading of cancellation tokens on some APIs is sufficient for the observed problem cases.

**Comment by KevinH-MS** [comment]
After a bit more thinking on this, I wonder if the answer isn't quite a bit simpler (at least for ```#r``` and ```#load```).  Currently, ```CompilationOptions``` has a ```MetadataReferenceResolver``` and a ```SourceReferenceResolver```.  As @tmat suggested, we could modify those to accept a ```CancellationToken```.  Resolvers with appropriate cancellation can be specified either when creating a ```Compilation``` or using ```.WithOptions``` on an existing ```Compilation```.  The ```.WithOptions``` route will attempt to reuse as much state as possible from the original ```Compilation``` (including ```SyntaxTrees``` and ```DeclarationTable``` if we merge PR #4346).  If cancellation is signalled, we can catch the Exception (similar to the way we catch IOException today), and report a diagnostic on the ```#r``` or ```#load``` syntax.  This is consistent with ```#r``` or ```#load``` of a file that doesn't exist or couldn't be read from disk.

**Comment by gafter** [comment]
@KevinH-MS In your suggested approach, if you cancel some operation, the whole `Compilation` is broken from that point forward (because it would have error symbols constructed at the time the cancellation occurred). That would be pretty bad in situations where the `Compilation` is used concurrently.

I suggest instead we consider adding an API
	
`void Compilation.ResolveReferencesAsync(CancellationToken)`

This API would not make any observable changes to the compilation other than resolving the references so that the compilation won’t have to block later. If you use this API and it completes normally, the `Compilation` would not block later on IO in the resolver. If you don't use this API or you let it get cancelled, the `Compilation` might block later on IO in the resolver.

**Comment by KevinH-MS** [comment]
What if Compilations forked after cancellation had occurred reset the resolved state (so that ```#r``` and ```#load``` would be re-resolved lazily in the forked Compilation)?

It doesn't seem terrible to me if some Compilations have incomplete information from syntax or metadata (error symbols).  It's certainly no worse than an error reading from disk, which the compiler handles (at least somewhat) gracefully.

I just worry that:

1.  It's undue burden on the consumer of the API to know they must call ```Compilation.ResolveReferencesAsync``` in order to get ***correct*** semantics.
2.  We're losing any benefits of laziness in the Compilation.

**Comment by gafter** [comment]
It is not necessary to call my proposed API to get ***correct*** semantics. The semantics are not affected by whether or not you call the API and whether the API is cancelled or not. That is why I wrote

> It would be completely correct to not use this API

In my proposal, the only thing that would no longer be lazy is resolving the #r and #load references from disk. Everything else would remain lazy as today. Unlike in your proposal, the semantics of the APIs would not be affected by what other consumers elect to do on other threads.

**Comment by KevinH-MS** [comment]
@gafter sorry, I didn't fully process the last paragraph in your proposal (I confused it with the proposal to do nothing with ```#r```/```#load``` if the "Resolve*Async" method wasn't called).

I still worry that throwing out laziness may lead to issues, but I guess we'd have to prototype and measure to know if that's an issue in practice.

**Comment by VSadov** [comment]
I agree with @gafter - Compilation.ResolveReferencesAsync(CancellationToken) could be an attractive solution.

Currently many compiler APIs assume that they are computationally bound and therefore they are not asynchronous.  For example CompileAsync does not make a lot of sense because Compile is predominantly compute-bound and is unlikely to yield the thread. At best you can have concurrency, but not asynchrony (in the overlapped operations sense).

If, however, we can isolate IO-bound operations, I can see how it would be useful to expose them in an Async and cancellable forms so that caller could “prep” the compilation. That is – if such operations can be isolated from the rest. Like in the proposed  Compilation.ResolveReferencesAsync(CancellationToken)

**Comment by AdamSpeight2008** [comment]
What about the case of aggressive \ complex optimisation that may take a lot longer, would it not make sense to make it cancellable?

**Comment by jasonmalinowski** [comment]
@AdamSpeight2008: emit already has cancellation in the API (although I don't know how far through it's threaded.) This is more a question what to do about the APIs where that's not easily doable.

**Comment by gafter** [comment]
@CyrusNajmabadi I'd like to bring this discussion to closure. Do you think opening an issue proposing my the API I suggested would be a sufficient resolution?

**Comment by GSPP** [comment]
What is the strategy to deal with APIs that fundamentally are not cancellable? For example I think the DNS resolution in HTTP request processing does not respond to cancellation and cannot be executed asynchronously. File IO is not cancellable at all in the current BCL. File IO might happen on a broken SMB share.

I don't think you can rely on cancellation actually taking effect when requested. If you want to offer a UI where the cancel button works you need to decouple physical cancellation from user-visible cancellation. The UI must be able to continue while uncancellable operations continue running and their result ignored and their side-effects isolated.

And if you have that kind of cancellation model then you don't really need physical cancellation anymore.

You might still want it to quickly shut down unnecessary memory, network and CPU consumption. But that is a secondary benefit.

**Comment by VSadov** [comment]
That is exactly how it works. Asynchrony is used for UI responsiveness and cancellation is used to limit resource waste.
If impact of consuming resources was completely isolated, cancellation would not be needed. Unfortunately most shared resources (memory, CPU, worker threads...) are not like that.

**Comment by GSPP** [comment]
So what about this:

> For the foreground thread this can provide a massively bad experience for the user

This seems to be fully handled then?

Regarding thread-pool: Is it really that easy to generate hundreds of threads? Is that a realistic case? Even if one keystroke causes one "hung" thread it would take a while to fill a limit that is in the hundreds.

Async IO is clearly a good solution but it is surprising how far you can go using threads. For example, all Roslyn actions prone to blocking could run on a separate thread-pool with 64KB stack size. That would expand the available thread count into the thousands.

I tested that Windows does not have a problem scheduling thousands of threads, especially when they are blocked almost all of the time. The machine became unresponsive only in the >20000 range.

Regarding pool starvation there are two practical solutions: Either raise the global pool limits or use a separate pool (a custom `TaskScheduler`).

I'm really not *recommending* using many threads but it is a viable approach. This approach often is underestimated in viability. I don't want to argue for any particular approach. Just trying to point out these observations in case they are helpful.

**Comment by jaredpar** [comment]
Closing after some discussion.
